### PR TITLE
Refactor metadata into new class MetadataElement

### PR DESCRIPTION
**Pull request by kswanson33**
related to openstax/cnx-recipes#3337

### CODE REVIEW

File: lib/kitchen/directions/bake_index/v1.rb [hunk 1]
```diff
@@ -1,142 +1,149 @@
...
-        @id = id
-        @group_by = group_by
-        @page_title = page_title
-      end
-    end
+# frozen_string_literal: true
+
+module Kitchen
+  module Directions
+    module BakeIndex
```
Comment (jpslav) [inline lib/kitchen/directions/bake_index/v1.rb:5]: I know Rubocop says to do it this way instead of `Kitchen::Directions::BakeIndex` but I'm not a huge fan b/c it loses like 10 characters for so much indenting.  What about ignoring this rule at least in directions which have a lot of nesting?

File: lib/kitchen/element_base.rb [hunk 2]
```diff
@@ -590,7 +592,8 @@ def clone
 
     # @!method pages
     #   Returns a pages enumerator
-    def_delegators :as_enumerator, :pages, :chapters, :terms, :figures, :notes, :tables, :examples
+    def_delegators :as_enumerator, :pages, :chapters, :terms, :figures, :notes, :tables, :examples,
+                   :metadatas # doing my best with the plural of data :/
```
Comment (jpslav) [inline lib/kitchen/element_base.rb:596]: that works, you can remove the comment

File: lib/kitchen/metadata_element.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            document: document,
+            enumerator_class: MetadataElementEnumerator,
+            short_type: :metadata)
+    end
+
+    # Returns set of selected data elements
+    #
+    # @return [ElementEnumerator]
+    #
+    def data
```
Comment (jpslav) [inline lib/kitchen/metadata_element.rb:23]: Is there a more meaningful name we can use?  Like `elements_to_keep`? (that name isn't awesome either but it conveys that these are child elements that we want to persist through baking.

File: spec/helpers/factory_helpers.rb [hunk 4]
```diff
@@ -65,5 +65,5 @@ def metadata_element
         <div data-type="subject" id="subject">Subject</div>
       </div>
     HTML
-  )
+  ).metadata
```
Comment (jpslav) [inline spec/helpers/factory_helpers.rb:68]: why this change?

**Review by jpslav — CHANGES_REQUESTED** [review]

File: spec/helpers/factory_helpers.rb (hunk 4 above)
Comment (kswanson33) [reply spec/helpers/factory_helpers.rb:68]: Specs fail without this change -- several functions (i.e. `BakeChapterGlossary.v1`)  have been updated to expect a `MetadataElement` instead of an `Element`. This function is called in the specs to generate metadata, so it must now also return a `MetadataElement`.

File: lib/kitchen/metadata_element.rb (hunk 3 above)
Comment (kswanson33) [reply lib/kitchen/metadata_element.rb:23]: What about these?: `data_to_keep` / `children_to_keep` / `descendants_to_keep`

File: spec/helpers/factory_helpers.rb (hunk 4 above)
Comment (jpslav) [reply spec/helpers/factory_helpers.rb:68]: ah right

**Review by jpslav — APPROVED** [review]
